src/Externals/ORM/EntityManager.php

```diff
@@ -0,0 +1,157 @@
+<?php
+declare(strict_types=1);
+
+namespace LoyaltyCorp\Multitenancy\Externals\ORM;
+
+use Doctrine\ORM\EntityManagerInterface as DoctrineEntityManager;
+use EoneoPay\Externals\ORM\Interfaces\EntityInterface;
+use LoyaltyCorp\Multitenancy\Database\Entities\Provider;
+use LoyaltyCorp\Multitenancy\Database\Exceptions\InvalidEntityOwnershipException;
+use LoyaltyCorp\Multitenancy\Database\Traits\HasProvider;
+use LoyaltyCorp\Multitenancy\Externals\Interfaces\ORM\EntityManagerInterface;
+
+final class EntityManager implements EntityManagerInterface
+{
+    /**
+     * Doctrine entity manager
+     *
+     * @var \Doctrine\ORM\EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * Create an internal entity manager
+     *
+     * @param \Doctrine\ORM\EntityManagerInterface $entityManager
+     */
+    public function __construct(DoctrineEntityManager $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \LoyaltyCorp\Multitenancy\Database\Exceptions\InvalidEntityOwnershipException If provider mismatch
+     */
+    public function flush(int $providerId): void
+    {
+        // Ensure that all entities pending write are for the correct provider
+        $unitOfWork = $this->entityManager->getUnitOfWork();
+
+        // Get entity changes
+        $changes = \array_merge(... [
```

Comment:
This wont catch any entities added to the unit of work during Doctrine event listeners - will we ever actually have event listeners modifying application entities?

I think that any doctrine listeners will generally be modifying library entities that wont have a `HasProvider` trait so it might not be a big deal?

Reply:
@merk it also doesn't catch collection changes where you're explicitly using `cascade={"persist"}`, I couldn't find a sure-fire way to do this and still have the provider available since the entity manager passed to a listener may not be the one we're performing the flush with (it's usually the doctrine version, not the wrapper).

Reply:
The only other option would be to register a doctrine subscriber that runs on preFlush, that we would hold a reference to inside "our" Entity manager.

When we call our protected flush, you could call a setter on the subscriber that adds the provider to be used for validation?